File: code/ex446.java
// DP

import java.util.*;

class Solution {
    public int numberOfArithmeticSlices(int[] A) {
        int res = 0;
        Map<Integer, Integer>[] dp = new HashMap[A.length];
        
        for (int i = 0; i < A.length; i++) {
            dp[i] = new HashMap<>();
            
            for (int j = 0; j < i; j++) {
                long d = (long)A[i] - (long)A[j];
                if (d > Integer.MAX_VALUE || d < Integer.MIN_VALUE)    continue;
                int diff = (int)d;
                
                int dj = dp[j].getOrDefault(diff, 0);
                int di = dp[i].getOrDefault(diff, 0);
                
                res += dj;
                dp[i].put(diff, di+ dj+ 1);
            }
        }
        
        return res;
    }
}

File: code/ex446.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
        int n = A.size();
        
        int cnt = 0;
        vector<unordered_map<int, int>> dp(n, unordered_map<int, int>());
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                long long d = (long long)A[i] - (long long)A[j];
                if (d >= INT_MAX || d <= INT_MIN)    continue;
                int diff = (int)d;
                
                if (dp[j].find(diff) != dp[j].end()) {
                    cnt += dp[j][diff];
                    dp[i][diff] += dp[j][diff] + 1;
                } else
                    dp[i][diff]++;
            }
        }
        
        return cnt;
    }
};